sway/sway.5.scd

```diff
@@ -186,21 +186,29 @@ set|plus|minus <amount>
 	Cycles the layout mode of the focused container through a list of layouts.
 
 *max_render_time* off|<msec>
-	Works together with *output max_render_time* to reduce the latency even
-	further by delaying the frame callbacks sent to a surface. When set to
-	a positive number of milliseconds, delays the frame callback in such a
-	way that the surface has the specified number of milliseconds to render
-	and commit new contents before being sampled by the compositor for the
-	next presentation. See *max_render_time* in *sway-output*(5) for
-	further details.
+	Controls when the relevant application is told to render this window, as a
+	positive number of milliseconds before the next time sway composites the
+	output. A smaller number leads to fresher rendered frames being composited
+	by sway and lower perceived input latency, but if set too low, the
+	application may not finish rendering before sway composites the output,
+	leading to delayed frames.
+
+	When set to off, the relevant application is told to render this window
+	immediately after display refresh. How much time is left for rendering
+	before sway composites the output at that point depends on the output
+	*max_render_time* setting.
 
 	To set this up for optimal latency:
-	. Set up *output max_render_time*.
+	. Set up *output max_render_time* (see *sway-output*(5)).
 	. Put the target application in _full-screen_ and have it continuously
 	  render something.
 	. Start by setting *max_render_time 1*. If the application drops
 	  frames, increment by *1*.
 
+	This setting only has an effect if a per-output *max_render_time* is in
```

Comment:
Applications need to support the presentation time protocol for this setting to work, right? Adding this to applications without presentation time support will do nothing?

Reply:
No, this affects frame callbacks. The presentation time protocol informs you when your frame ended up being shown on screen, and is unaffected.

(The previous docs noted that the *backend* needs to have some form of presentation feedback and ability to estimate next frame time for any of this to work at all, but I simplified that to just say that it only works for wayland and drm.)

Reply:
So do applications like mpv that use presentation time benefit from more configurable frame callbacks?

Reply:
Frame callbacks are not configurable from the application perspective. Either it asks for them or it doesn't. You can use per-view `max_render_time` to control when sway will send them.

An application like mpv could use presentation time to know how long it took after it submitted its frame until it actually ended up on screen. Then on next frame callback, it could compensate for the delay by sending video frames from further ahead in the video, so everything lines up correctly when display finally shows the content.

(It is also possible to not use frame callbacks at all, but using only presentation time would likely lead to delayed frames as it does not expose the composition time, only presentation time.)